File: src/Ecr.Domain/Entities/Configuration/ValidationRule.cs
using Ecr.Domain.Abstractions;
using Ecr.Domain.Enums;
using Ecr.Domain.ValueObjects;

namespace Ecr.Domain.Entities.Configuration;

/// <summary>Правило валідації. Рівень визначає, чи блокує воно запис (R-B3).</summary>
public sealed class ValidationRule : Entity<int>
{
    private ValidationRule() { }

    /// <param name="tableDefId">Таблиця, якій належить правило.</param>
    /// <param name="code">Код правила; він же в повідомленні порушення.</param>
    /// <param name="severity">Рівень: <c>Info</c>, <c>Warning</c>, <c>Error</c>.</param>
    /// <param name="scope">0 Cell, 1 Row, 2 Table, 3 Document.</param>
    /// <param name="expression">Предикат нашою мовою.</param>
    /// <param name="message">Текст порушення мовами каталогу.</param>
    /// <param name="columnDefId">
    /// Колонка, до якої прив'язане правило; <c>null</c> — до всіх колонок
    /// таблиці.
    /// </param>
    /// <remarks>
    /// ⚠ <paramref name="columnDefId"/> зʼявився з аудиту: властивість
    /// існувала, рушій валідації її читав — і не було
    /// жодного способу її задати. Тобто гілка «правило лише для цієї
    /// колонки» була недосяжна, і кожне коміркове правило застосовувалося
    /// до всієї таблиці.
    /// </remarks>
    public ValidationRule(int tableDefId, EcrCode code, ValidationSeverity severity, byte scope,
                          string expression, LocalizedText message, int? columnDefId = null)
    {
        TableDefId = tableDefId;
        Code = code.Value;
        Severity = severity;
        Scope = scope;
        ColumnDefId = columnDefId;
        Expression = expression;
        MessageL10n = message;
        IsActive = true;
    }

    public int TableDefId { get; private set; }
    public string Code { get; private set; } = null!;
    public ValidationSeverity Severity { get; private set; }

    /// <summary>0 Cell, 1 Row, 2 Table, 3 Document.</summary>
    public byte Scope { get; private set; }

    public int? ColumnDefId { get; private set; }
    public string Expression { get; private set; } = null!;
    public LocalizedText MessageL10n { get; private set; } = null!;
    public bool IsActive { get; private set; }

    /// <summary>
    /// Чи блокує правило збереження. Блокує **лише** комірковий <c>Error</c>:
    /// заборона зберегти проміжний стан робить роботу з великою таблицею
    /// неможливою (R-B3).
    /// </summary>
    public bool BlocksSave => Severity == ValidationSeverity.Error && Scope == 0;

    /// <summary>Перезаписує налаштування правила.</summary>
    /// <remarks>
    /// ⛔ Сеттер потрібен для <c>PUT …/validation-rules/{code}</c>
    /// (авторство структури шаблону, W5.4, продовження зрізу ФВ-2.1..ФВ-2.5
    /// на <c>ValidationRule</c>): до цього кожне поле, крім <see cref="Code"/>
    /// і <see cref="TableDefId"/> (адреса й батько — не змінюються), задавав
    /// лише конструктор, і повторний виклик тим самим кодом не мав як
    /// оновити наявне правило.
    ///
    /// ⚠ <see cref="Code"/> і <see cref="TableDefId"/> тут НЕ змінюються
    /// навмисно — так само, як <see cref="SheetDef.Code"/> в
    /// <see cref="SheetDef.Rename"/>: код — це ідентичність і адреса в API,
    /// а таблиця-батько — те, через що правило взагалі знайшли.
    /// </remarks>
    /// <param name="severity">Рівень: <c>Info</c>, <c>Warning</c>, <c>Error</c>.</param>
    /// <param name="scope">0 Cell, 1 Row, 2 Table, 3 Document.</param>
    /// <param name="expression">Предикат нашою мовою.</param>
    /// <param name="message">Текст порушення мовами каталогу.</param>
    /// <param name="columnDefId">Колонка, до якої прив'язане правило; <c>null</c> — до всіх колонок таблиці.</param>
    /// <param name="isActive">Чи діє правило.</param>
    public void Update(
        ValidationSeverity severity, byte scope, string expression, LocalizedText message,
        int? columnDefId, bool isActive)
    {
        ArgumentNullException.ThrowIfNull(message);
        ArgumentException.ThrowIfNullOrWhiteSpace(expression);

        Severity = severity;
        Scope = scope;
        ColumnDefId = columnDefId;
        Expression = expression;
        MessageL10n = message;
        IsActive = isActive;
    }
}

File: src/Ecr.Web/src/features/templates/validationRule.ts
import type { ValidationRuleDto, ValidationSeverity } from '@/api/types';
import type { LocalizedValue } from '@/shared/ui/LocalizedInput';

/**
 * Чернетка правила валідації таблиці в редакторі (W5.4, продовження
 * `ФВ-2.1`..`ФВ-2.5` на `ValidationRule`).
 *
 * ⚠ Окремий тип від `ValidationRuleDto` — та сама причина, що й у
 * `SheetDraft`: `id` і `tableDefId` рахує сервер, форма ними не керує.
 */
export interface ValidationRuleDraft {
  /** Код правила; після створення не змінюється — це його адреса в API. */
  readonly code: string;
  readonly severity: ValidationSeverity;
  readonly scope: number;
  readonly expression: string;
  readonly messageL10n: LocalizedValue;
  readonly columnDefId: number | null;
  readonly isActive: boolean;

  /** Чи це нова чернетка: код нової ще можна набрати. */
  readonly isNew: boolean;
}

/** Область дії правила: 0 Cell, 1 Row, 2 Table, 3 Document. */
export const ValidationScopes: readonly { value: number; label: string }[] = [
  { value: 0, label: 'Cell' },
  { value: 1, label: 'Row' },
  { value: 2, label: 'Table' },
  { value: 3, label: 'Document' },
];

/** Порожня чернетка нового правила. */
export function emptyValidationRuleDraft(): ValidationRuleDraft {
  return {
    code: '',
    severity: 'Error',
    scope: 0,
    expression: '',
    messageL10n: {},
    columnDefId: null,
    isActive: true,
    isNew: true,
  };
}

/** Чернетка з наявного правила — для правки. */
export function validationRuleDraftOf(rule: ValidationRuleDto): ValidationRuleDraft {
  return {
    code: rule.code,
    severity: rule.severity,
    scope: rule.scope,
    expression: rule.expression,
    messageL10n: { ...(rule.messageL10n.values ?? {}) },
    columnDefId: rule.columnDefId,
    isActive: rule.isActive,
    isNew: false,
  };
}

/** Що саме заважає зберегти чернетку. */
export type ValidationRuleBlocker = 'Code' | 'Expression' | 'Message';

/**
 * Чому чернетку ще не можна зберегти; `null` — можна.
 *
 * ⛔ Не копія серверних правил: сервер відхиляє недопустимий код
 * `ECR-CFG-0422` сам, форма лише не везе в мережу те, що напевно повернеться
 * відмовою.
 */
export function whyCannotSaveValidationRule(draft: ValidationRuleDraft): ValidationRuleBlocker | null {
  if (draft.code.trim().length === 0) return 'Code';
  if (!/^[A-Za-z][A-Za-z0-9_]{0,63}$/.test(draft.code)) return 'Code';
  if (draft.expression.trim().length === 0) return 'Expression';
  if (Object.values(draft.messageL10n).every((text) => text.trim().length === 0)) return 'Message';

  return null;
}
